solutions/usaco-833.mdx

```diff
@@ -329,189 +328,255 @@ this solution if you find it unfamiliar and come back later.
-		else if (ancestor.equals(X) || ancestor.equals(Y)) { 
-			out.print(String.format("%s is the ", ancestor));
-			for (int i = 0; i < (ancestor.equals(X) ? minYDist : minXDist) - 2; i++)
-				out.print("great-");
-			if ((ancestor.equals(X) ? minYDist : minXDist) > 1)
-				out.print("grand-");
-			out.println(String.format("mother of %s", ancestor.equals(X) ? Y : X));
+		PrintWriter written = new PrintWriter("family.out");
+		if (ancestor == null) {
+			written.println("NOT RELATED");
+		} else if (ancestor.equals(cows[0]) || ancestor.equals(cows[1])) {
+			written.printf("%s is the ", ancestor);
+			int dist = ancestor.equals(cows[0]) ? dist1 : dist0;
+			for (int i = 0; i < dist - 2; i++) {
+				written.print("great-");
+			}
+			if (dist > 1) {
+				written.print("grand-");
+			}
+			String young = ancestor.equals(cows[0]) ? cows[1] : cows[0];
+			written.printf("mother of %s", young);
 		} else {
-			if (minXDist == 1 || minYDist == 1) {
-				// if X and Y both have a distance 1 to their ancestor,
-				// then the ancestor is their mother, so they are siblings
-				if (minXDist == 1 && minYDist == 1)
-					out.println("SIBLINGS");
-				// if the distance between the ancestor and X or Y is only 1,
-				// then it's an aunt relationship
-				else {
-					boolean auntIsX = minXDist == 1;
-					out.print(String.format("%s is the ", auntIsX ? X : Y));
-					for (int i = 0; i < (auntIsX ? minYDist : minXDist) - 2; i++)
-						out.print("great-");
-					out.println(String.format("aunt of %s", auntIsX ? Y : X));
+			if (dist0 == 1 || dist1 == 1) {
+				if (dist0 == 1 && dist1 == 1) {
+					written.println("SIBLINGS");
+				} else {
+					boolean auntIs0 = dist0 == 1;
+					written.printf("%s is the ", auntIs0 ? cows[0] : cows[1]);
+					for (int i = 0; i < (auntIs0 ? dist1 : dist0) - 2; i++) {
+						written.print("great-");
+					}
+					written.printf("aunt of %s", auntIs0 ? cows[1] : cows[0]);
 				}
-				// otherwise, they are cousins as defined in the question
-			} else
-				out.println("COUSINS");
+			} else {
+				written.println("COUSINS");
+			}
 		}
-		out.close();
+		written.close();
 	}
 }
 ```
 
 </JavaSection>
-
-
 <CPPSection>
 
 ```cpp
-CodeSnip{CPP Short Template}
+#include <bits/stdc++.h>
+using namespace std;
 
 int main() {
-	setIO("family");
+	freopen("family.in", "r", stdin);
 	int N;
-	string c[3];
-	cin >> N >> c[1] >> c[2];
-	map<string,int> dist[3];
-	dist[1][c[1]] = 0;
-	dist[2][c[2]] = 0;
-
-	map<string,string> edge; //edge list
+	vector<string> cows(2);
+	cin >> N >> cows[0] >> cows[1];
 
+	map<string, string> edge;
 	for (int i = 0; i < N; ++i) {
 		string u, v;
 		cin >> v >> u;
 		// If v is a mother of u, then there is an edge from node u to node v.
 		edge[u] = v;
 	}
 
-	for (int k = 1; k <= 2; ++k) {
-		// BFS on directed tree
-		// to find the ancestors of both cows and the distance to them
-		string node = c[k];
+	vector<map<string, int>> dist(2);
```

Comment:
how does this work? what does the dimensions represent?

Reply:
vector of all the ancestors & their distance

Reply:
make a comment for all the languages 💀

Reply:
i did you bozo